crates/sui-core/src/consensus_adapter.rs

```diff
@@ -456,18 +520,81 @@ impl ConsensusAdapter {
     }
 }
 
-/// Returns a position of the current validator in ordered list of validator to submit transaction
-pub fn position_submit_certificate(
+impl CheckConnection for ConnectionMonitorStatus {
+    fn check_connection(
+        &self,
+        ourself: &AuthorityName,
+        authority: &AuthorityName,
+    ) -> Option<ConnectionStatus> {
+        if ourself == authority {
+            return Some(ConnectionStatus::Connected);
+        }
+
+        let mapping = self.authority_names_to_peer_ids.load_full();
+        let peer_id = match mapping.get(authority) {
+            Some(p) => p,
+            None => {
+                warn!(
+                    "failed to find peer {:?} in connection monitor listener",
+                    authority
+                );
+                return None;
+            }
+        };
+
+        let res = match self.connection_statuses.try_get(peer_id) {
+            TryResult::Present(c) => Some(c.value().clone()),
+            TryResult::Absent => None,
+            TryResult::Locked => {
+                // update is in progress, assume the status is still or becoming disconnected
+                Some(ConnectionStatus::Disconnected)
+            }
+        };
+        res
+    }
+    fn update_mapping_for_epoch(
+        &self,
+        authority_names_to_peer_ids: HashMap<AuthorityName, PeerId>,
+    ) {
+        self.authority_names_to_peer_ids
+            .swap(Arc::new(authority_names_to_peer_ids));
+    }
+}
+
+impl CheckConnection for ConnectionMonitorStatusForTests {
+    fn check_connection(
+        &self,
+        _ourself: &AuthorityName,
+        _authority: &AuthorityName,
+    ) -> Option<ConnectionStatus> {
+        Some(ConnectionStatus::Connected)
+    }
+    fn update_mapping_for_epoch(
+        &self,
+        _authority_names_to_peer_ids: HashMap<AuthorityName, PeerId>,
+    ) {
+    }
+}
+
+pub fn order_validators_for_submission(
     committee: &Committee,
-    ourselves: &AuthorityName,
     tx_digest: &TransactionDigest,
-) -> usize {
+) -> Vec<AuthorityName> {
     // the 32 is as requirement of the default StdRng::from_seed choice
     let digest_bytes = tx_digest.into_inner();
 
     // permute the validators deterministically, based on the digest
     let mut rng = StdRng::from_seed(digest_bytes);
-    let validators = committee.shuffle_by_stake_with_rng(None, None, &mut rng);
+    committee.shuffle_by_stake_with_rng(None, None, &mut rng)
+}
+
+/// Returns a position of the current validator in ordered list of validator to submit transaction
+pub fn position_submit_certificate(
```

Comment:
Is this function used anywhere outside of tests? If no, we should remove it (and update tests to use some other function that we made perhaps?)

Reply:
This was and still is testing the order_validators_for_sumbmission function, I've moved its location to become clearly a test helper function

Reply:
But this is exactly my concern - we used to use `position_submit_certificate` in the code, and we wrote unit tests for it.

Now you switched to using `submission_position` function in the prod tests, but unit test continue to use `position_submit_certificate`. Don't you think unit tests also need to be updated to test same function(`submission_position`) that production code uses??

Reply:
Ok, so you are asking for a unit test for the function `submission_possiotion`, or more specifically `check_submission_wrt_connectivity`? I still see the random ordering of validators, or rather the functionality in `order_validators_for_submission` to be the part that needs testing, since without that part working as expected, we may have no-one submit a tx.  The functionality of removing authorities from a list doesn't warrant a unit test, in my opinion, so I didn't add one. Also I didn't add a unit test for checking connectivity, because we rely on the connectivity monitor in our dashboards, I did assume whoever created those has verified their correctness.

Reply:
Is it hard to switch unit test to use `submission_position` and pass your `ConnectivityStatusForTest` to consensus adapter?
The reason we prefer to have unit tests for the functions that are actually used in real code is that when something change in those functions, we want to get signal from unit tests.